### PR TITLE
lz4: use CMake to build

### PR DESCRIPTION
**Pull request by Scrumplex**
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/318509

This packages CMake configuration files. It also simplifies the derivation quite a bit.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] pkgsCross.mingwW64
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Scrumplex** [comment]
Both `nix build -vL .#pkgsCross.mingwW64.lz4` and `nix build -vL .#lz4` build fine for me

**Comment by szlend** [comment]
I'd like to test that this still cross-compiles from Darwin to Linux as soon as I find the time.

File: pkgs/tools/compression/lz4/default.nix [hunk 1]
```diff
@@ -1,19 +1,29 @@
...
+    owner = "lz4";
   };
 
+  nativeBuildInputs = [
+    cmake
+  ];
+
+  buildInputs = lib.optional finalAttrs.doCheck valgrind;
+
+  # TODO(@Ericson2314): Separate binaries and libraries
```
Comment (marsam) [inline pkgs/tools/compression/lz4/default.nix:24]: I believe this comment is outdated(?)
```suggestion
```

File: pkgs/tools/compression/lz4/default.nix [hunk 2]
```diff
@@ -22,38 +32,17 @@ stdenv.mkDerivation rec {
...
-  # TODO(@Ericson2314): Separate binaries and libraries
-  outputs = [ "bin" "out" "dev" ];
-
-  buildInputs = lib.optional doCheck valgrind;
-
-  enableParallelBuilding = true;
+  cmakeFlags = [
+    (lib.cmakeBool "BUILD_SHARED_LIBS" enableShared)
+    (lib.cmakeBool "BUILD_STATIC_LIBS" enableStatic)
+  ];
```
Comment (marsam) [inline pkgs/tools/compression/lz4/default.nix:38]: I think these aren't needed. By default it builds the shared lib, and `pkgsStatic` adds the cmakeFlag to build the static one
```suggestion
```

File: pkgs/tools/compression/lz4/default.nix [hunk 3]
```diff
@@ -22,38 +32,17 @@ stdenv.mkDerivation rec {
...
-  ]
-    # TODO make full dictionary
-    ++ lib.optional stdenv.hostPlatform.isMinGW "TARGET_OS=MINGW"
-    ++ lib.optional stdenv.hostPlatform.isLinux "TARGET_OS=Linux"
-    ;
+  cmakeDir = "../build/cmake";
+  cmakeBuildDir = "build-dist";
 
   doCheck = false; # tests take a very long time
   checkTarget = "test";
```
Comment (marsam) [inline pkgs/tools/compression/lz4/default.nix:45]: I think we could add:
```suggestion
  passthru.tests = {
    version = testers.testVersion {
      package = finalAttrs.finalPackage;
      version = "v${finalAttrs.version}";
    };
    pkg-config = testers.hasPkgConfigModules {
      package = finalAttrs.finalPackage;
      moduleNames = [ "liblz4" ];
    };
  };
```

File: pkgs/tools/compression/lz4/default.nix [hunk 4]
```diff
@@ -67,4 +56,4 @@ stdenv.mkDerivation rec {
     license = with licenses; [ bsd2 gpl2Plus ];
     platforms = platforms.all;
```
Comment (marsam) [inline pkgs/tools/compression/lz4/default.nix:57]:
```suggestion
    platforms = platforms.all;
    mainProgram = "lz4";
```

File: pkgs/tools/compression/lz4/default.nix [hunk 5]
```diff
@@ -1,19 +1,29 @@
...
   };
 
+  nativeBuildInputs = [
+    cmake
+  ];
+
+  buildInputs = lib.optional finalAttrs.doCheck valgrind;
+
+  # TODO(@Ericson2314): Separate binaries and libraries
+  outputs = [ "bin" "out" "dev" ];
```
Comment (marsam) [inline pkgs/tools/compression/lz4/default.nix:25]: nitpick
```suggestion
  outputs = [ "bin" "out" "dev" "man" ];
```

**Comment by szlend** [comment]
I can confirm cross-compilation from `aarch64-darwin` to `aarch64-linux` still works.

**Review by getchoo — APPROVED** [review]
changes LGTM. i was also able to build on x86_64-linux, aarch64-linux, and cross compile to mingwW64 from both

**Comment by makuru-dd** [comment]
This pull request seems to fix problems I have with cmake not finding `lz4`.

**Review by makuru-org — APPROVED** [review]

**Review by h7x4 — APPROVED** [review]

**Comment by mweinelt** [comment]
This breaks at least `arrow-cpp`, which now fails to find lz4.

Likely commit is ead868553bb3e9b7ef6834843050299e297dd746, that fails to evaluate, because `finalAttrs` gets only introduced in 59238b728ca666478123c95e591aff727d0ae141.

**Pull request merged**